Added websocket support to GraphQL using the graphql-ws protocol

spacecowboy opened the pull request:
Motivation:

WebSocket support was recently added to Armeria but it lacks built-in support for GraphQL subscriptions over WebSockets. This adds that functionality.

Modifications:

- Modified `GraphqlBuilder` so that websocket support can be enabled or disabled
- Added WebSocket marker interface so that any service can be allowed to support it and not just `WebSocketService`

Result:

- Closes #4999
- Optional WebSocket support is added for GraphQL.
- Some (Unstable) API surfaces might change.

<!--
Visit this URL to learn more about how to write a pull request description:
https://armeria.dev/community/developer-guide#how-to-write-pull-request-description
-->

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/line/armeria?pullRequest=5280) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/line/armeria?pullRequest=5280) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/line/armeria?pullRequest=5280) it.</sub>

spacecowboy commented:
Still a work in progress. I have successfully opened a web socket and executed a subscription through it, but I'm currently getting a Null-problem with the ExecutionResult.

- Not sure what to do with ServiceRequestContext since a websocket doesn't have a request

spacecowboy commented:
In the latest commit [abdb78e](https://github.com/line/armeria/pull/5280/commits/abdb78e8a98785f69dba92287bc19068bd510d78) I have successfully executed a subscription.

![Screenshot at 2023-11-06 14-11-42](https://github.com/line/armeria/assets/223655/bf9f9dc5-00fe-460c-b215-cbd2181653ba)

jrhee17 commented:
Nice! Let us know when this is ready for review 🙇

spacecowboy commented:
@jrhee17 @minwoox I think it's ready for review now. I think I've gotten as far as I can without feedback. Would appreciate some tips on how to write some good tests for this.

jrhee17 commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/DefaultGraphqlService.java:
```diff
@@ -30,49 +31,86 @@
+     * Used for web sockets where there is no request context.
+     */
+    @Nullable
+    private Supplier<? extends DataLoaderRegistry> dataLoaderRegistrySupplier;
```
A `GraphQlService` is shared by multiple connections/requests/event loops so I'm concerned that this field will suffer from race conditions (a registry for requestA may be used for requestB)

What do you think of adding the `DefaultClientRequestContext`  to the `executeGraphql` method parameter and using the provided `dataLoaderRegistryFunction` instead?

e.g.
```java
 interface GraphqlExecutor {
-    ExecutionResult executeGraphql(ExecutionInput.Builder builder);
+    ExecutionResult executeGraphql(Builder builder, ServiceRequestContext ctx);
```

```java
+    private final ServiceRequestContext ctx;
 
     GraphqlWSSubProtocol(ServiceRequestContext ctx, GraphqlExecutor executor) {
         upgradeCtx = GraphqlServiceContexts.graphqlContext(ctx);
         graphqlExecutor = executor;
+        this.ctx = ctx;
     }
```

```java
final ExecutionResult executionResult = graphqlExecutor.executeGraphql(executionInput, ctx);
```

spacecowboy replied:
Functionally equivalent to what I did with the provider I think but if you think it's clearer then sure I can change it.

Note that the supplier was defined as such:

```
protected HttpResponse doWebSocketUpgrade(ServiceRequestContext ctx, HttpRequest req) throws Exception {
        if (webSocketService != null) {
            dataLoaderRegistrySupplier = () -> dataLoaderRegistryFunction.apply(ctx);
```

Which used the `dataLoaderRegistryFunction` with the same `ctx` as was used to establish the socket, just like your version does.

jrhee17 replied:
Rather than clarity, I was worried about race conditions since `doWebSocketUpgrade` can be called concurrently by different requests.

i.e.
- RequestA sets `dataLoaderRegistrySupplier` with ctxA
- RequestB sets `dataLoaderRegistrySupplier` with ctxB
- Later, RequestA uses ctxB

jrhee17 commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/DefaultGraphqlService.java:
```diff
@@ -92,38 +130,62 @@ protected HttpResponse executeGraphql(ServiceRequestContext ctx, GraphqlRequest
+            .executionId(ExecutionId.generate())
+            .build();
+
+        return graphQL.execute(executionInput);
```
This looks like a blocking call on the event loop.
What do you think of reusing the logic at L164 `(DefaultGraphqlService#execute(ServiceRequestContext ctx, ExecutionInput input, MediaType produceType))` to return a future instead?

spacecowboy replied:
Still need to consume the future in `GraphqlWSSubProtocol`. That would still block the event loop. Or could that use the `BlockingTaskExecutor` to establish the subscriber?

jrhee17 reviewed (commented):
Looks good overall! Left some minor suggestions 🙇 

> Would appreciate some tips on how to write some good tests for this.

Basically, I imagine an integration test that looks similar to `GraphqlServiceTest` which involves setting up a server and verifying requests correctly reach the server.

The problem is we don't have a `graphql-ws` client, so I guess you will either need to
- find a third party library (I wasn't able to find any)
- just manually send over `graphql-ws` messages manually via armeria's `WebSocketClient`

spacecowboy commented on core/src/main/java/com/linecorp/armeria/server/Http1RequestDecoder.java:
```diff
@@ -167,7 +167,7 @@ public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception
 
                     final String path = HttpHeaderUtil
-                            .maybeTransformAbsoluteUri(nettyReq.uri(), cfg.absoluteUriTransformer());
+                        .maybeTransformAbsoluteUri(nettyReq.uri(), cfg.absoluteUriTransformer());
```
Fixed.

spacecowboy commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/GraphqlWSSubProtocol.java:
```diff
@@ -0,0 +1,279 @@
+                    writeConnectionAck(out);
+                    break;
+                case "ping":
+                    ensureInitiated();
```
Reading it again, and I think it is actually allowed. changing..

spacecowboy commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/GraphqlWSSubProtocol.java:
```diff
@@ -0,0 +1,279 @@
+            final String id;
+
+            switch (type) {
+                case "connection_init":
```
Your right. Fixing...

spacecowboy commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/GraphqlWSSubProtocol.java:
```diff
@@ -0,0 +1,279 @@
+                            return;
+                        }
+
+                        final Publisher<ExecutionResult> publisher = executionResult.getData();
```
`subscribe`-event *should* only contain a graphql subscription-query. But that depends on how the client implementation handles faulty input I suppose. So should add a check.

spacecowboy commented:
> Basically, I imagine an integration test that looks similar to `GraphqlServiceTest` which involves setting up a server and verifying requests correctly reach the server.
> 
>     * just manually send over `graphql-ws` messages manually via armeria's `WebSocketClient`

I added  a basic happy path test as a starting point

jrhee17 commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/GraphqlServiceBuilder.java:
```diff
@@ -263,6 +265,14 @@ public GraphqlServiceBuilder useBlockingTaskExecutor(boolean useBlockingTaskExec
     }
 
+    /**
+     * Sets whether the service supports WebSockets for subscriptions.
```
What do you think of leaving a link so that people can refer to it?

```
Sets whether [[GraphQL over WebSocket Protocol](https://github.com/enisdenjo/graphql-ws/blob/master/PROTOCOL.md)] is supported.
```

spacecowboy replied:
Done

jrhee17 commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/GraphqlService.java:
```diff
@@ -42,6 +42,6 @@ static GraphqlServiceBuilder builder() {
     static GraphqlService of(GraphQL graphQL) {
         return new DefaultGraphqlService(graphQL, ctx -> new DataLoaderRegistry(), false,
-                                         GraphqlErrorHandler.of());
+                                         GraphqlErrorHandler.of(), true);
```
The default behavior of the builder and factory should match (either opt-in or opt-out by default).

I'm not strong on either default and think it can go either way.

```suggestion
                                         GraphqlErrorHandler.of(), false);
```

spacecowboy replied:
Done

jrhee17 commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/DefaultGraphqlService.java:
```diff
@@ -115,7 +173,7 @@ private HttpResponse execute(
                         return HttpResponse.ofJson(HttpStatus.NOT_IMPLEMENTED,
                                                    produceType,
-                                                   toSpecification("WebSocket is not implemented"));
+                                                   toSpecification("Use websocket for subscription"));
```
```suggestion
                                                   toSpecification("Use GraphQL over WebSocket for subscription"));
```

spacecowboy replied:
Done

jrhee17 commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/GraphqlSubscriber.java:
```diff
@@ -0,0 +1,112 @@
+import graphql.ExecutionResult;
+
+@SuppressWarnings("ReactiveStreamsSubscriberImplementation")
+class GraphqlSubscriber implements Subscriber<ExecutionResult> {
```
Naming) What do you think of `ExecutionResultSubscriber`?

```suggestion
class ExecutionResultSubscriber implements Subscriber<ExecutionResult> {
```

spacecowboy replied:
I think it's better

jrhee17 commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/GraphqlSubscriber.java:
```diff
@@ -0,0 +1,112 @@
+    }
+
+    @Override
+    public void onError(Throwable t) {
```
Question) Is there no need to send a `Error` message and remove this subscriber from `graphqlSubscriptions`?

spacecowboy replied:
I've added it

jrhee17 commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/GraphqlSubscriber.java:
```diff
@@ -0,0 +1,112 @@
+    }
+
+    @Override
+    public void onComplete() {
```
Question) Is there no need to send a "Complete" message from server-side when `onComplete` is invoked? (and also remove the subscriber from the `graphqlSubscriptions` to save memory?)

> Server dispatches the Complete message indicating that the source stream has completed
> Client completes the stream observer 

https://github.com/enisdenjo/graphql-ws/blob/master/PROTOCOL.md#streaming-operation

spacecowboy replied:
I think it will only ever reach complete in the case that the client sends the COMPLETE message first. I'm not sure there is a scenario where the server completes the subscription

jrhee17 commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/GraphqlWebSocketSubscriber.java:
```diff
@@ -0,0 +1,121 @@
+            // PONG is a noop
+            case PONG:
+                // These below will never happen?
+            case BINARY:
+            case CONTINUATION:
```
Is there any reference on this in the graphql-ws spec?

If there isn't any specification:
- I feel like we can probably handle `BINARY` the same way that we handle `TEXT` by just using the `text()` method.
- For `CONTINUATION`, I guess it's not necessary to handle now and we can just log. I guess it would involve checking the `webSocketFrame.isFinalFragment()` flag and accumulating to a queue.

jrhee17 commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/GraphqlSubscriber.java:
```diff
@@ -0,0 +1,112 @@
+                protocol.sendGraphqlErrors(executionResult.getErrors());
+                if (subscription != null) {
+                    subscription.cancel();
+                }
```
Question) Is there no need to remove this subscriber from `graphqlSubscriptions`?

spacecowboy replied:
Maybe. I've added it

jrhee17 reviewed (commented):
Sorry about the late review! Looks almost done 👍

spacecowboy commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/DefaultGraphqlService.java:
```diff
@@ -99,6 +130,33 @@ protected HttpResponse executeGraphql(ServiceRequestContext ctx, GraphqlRequest
+    public CompletableFuture<ExecutionResult> executeGraphql(ServiceRequestContext ctx, Builder builder) {
+        final ExecutionInput executionInput = builder
+                .dataLoaderRegistry(dataLoaderRegistryFunction.apply(ctx))
+                .executionId(ExecutionId.generate())
```
From what I found, the executionId is set based on the RequestContext.

And operations inside the websockets don't have a request context.

The context they do have is the initial subscribe request, and we don't want every operation to have the same ID.

spacecowboy replied:
I could but it seems a bit too broad to allow BlockingCalls inside the entire method.

But where in CI is this an issue? I'm not seeing a failure related to this, and is generating a UUID really considered a blocking call?

spacecowboy replied:
Added a block hound integration file

spacecowboy commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/GraphqlWSSubProtocol.java:
```diff
@@ -0,0 +1,294 @@
+                        final CompletableFuture<ExecutionResult> future =
+                                graphqlExecutor.executeGraphql(ctx, executionInput);
+
+                        final ExecutionResult executionResult = future.get();
```
Rewritten according to suggestion.

spacecowboy commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/GraphqlWSSubProtocol.java:
```diff
@@ -0,0 +1,294 @@
+                            throw new Exception("Result of operation was not a subscription");
+                        }
+
+                        final Publisher<ExecutionResult> publisher = executionResult.getData();
```
Done

spacecowboy commented on core/src/main/java/com/linecorp/armeria/server/websocket/WebSocketService.java:
```diff
@@ -61,7 +61,7 @@
  */
 @UnstableApi
-public final class WebSocketService extends AbstractHttpService {
+public final class WebSocketService extends AbstractHttpService implements IWebSocketService {
```
Good suggestion. Done

spacecowboy commented:
Updated @jrhee17 . The only open question for me is about the ExecutionID. Typically the executionId is generated from the RequestContext, and there is no RequestContext inside a websocket. Only the initial RequestContext from the websocket upgrade request is available, but that would generate the same ID which is not desirable.

jrhee17 commented on graphql/src/test/java/com/linecorp/armeria/server/graphql/GraphqlServiceSubscriptionTest.java:
```diff
@@ -78,14 +89,57 @@ void testSubscription() {
+    @Test
+    void testSubscriptionOverWebSocket() {
+        final WebSocketClient webSocketClient =
+                WebSocketClient.builder(server.uri(SessionProtocol.H1C, SerializationFormat.WS))
```
Do you mind testing for both HTTP1/2 for completeness?

In order to enable HTTP2, I think you will need to modify `AbstractGraphqlService#exchangeType` so that `BIDI_STREAMING` is returned when websockets are enabled.

jrhee17 commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/GraphqlWSSubProtocol.java:
```diff
@@ -0,0 +1,356 @@
+                        final Map<String, Object> variables = toMapFromJson(payload.get("variables"));
+                        final Map<String, Object> extensions = toMapFromJson(payload.get("extensions"));
+
+                        final ExecutionInput.Builder executionInput =
+                                ExecutionInput.newExecutionInput()
+                                              .graphQLContext(upgradeCtx)
+                                              .graphQLContext(
+                                                      connectionCtx)
+                                              .query(query)
+                                              .variables(variables)
+                                              .operationName(
+                                                      operationName)
+                                              .extensions(extensions);
```
nit; do you mind reformatting these lines for aesthetics sake?

```
ExecutionInput.newExecutionInput()
              .graphQLContext(upgradeCtx)
              .graphQLContext(connectionCtx)
              .query(query)
              .variables(variables)
              .operationName(operationName)
              .extensions(extensions);
```

jrhee17 commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/GraphqlWSSubProtocol.java:
```diff
@@ -0,0 +1,356 @@
+                    @Override
+                    public void completeWithError(Throwable cause) {
+                        writeError(out, id, cause);
+                        writeComplete(out, id);
```
Question) if I understand correctly, we shouldn't send messages after the error message

> This message terminates the operation and no further messages will be sent.

ref: https://github.com/enisdenjo/graphql-ws/blob/master/PROTOCOL.md#error

```suggestion
```

jrhee17 commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/GraphqlWSSubProtocol.java:
```diff
@@ -0,0 +1,356 @@
+                });
+
+        graphqlSubscriptions.put(id, executionResultSubscriber);
+        streamMessage.subscribe(executionResultSubscriber);
```
So that we can guarantee we access the map data structure from a single thread (to prevent concurrent modification exceptions)

```suggestion
        streamMessage.subscribe(executionResultSubscriber, ctx.eventLoop());
```

jrhee17 commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/GraphqlWSSubProtocol.java:
```diff
@@ -0,0 +1,356 @@
+                        final CompletableFuture<ExecutionResult> future =
+                                graphqlExecutor.executeGraphql(ctx, executionInput);
+
+                        future.handle(((executionResult, throwable) -> {
+                            handleBlocking(out, id, executionResult, throwable);
+                            return null;
+                        }));
```
It isn't guaranteed which thread will complete the future, so `graphqlSubscriptions` can be handled from multiple threads (concurrent modification exception).

```suggestion
                        future.handleAsync(((executionResult, throwable) -> {
                            handleBlocking(out, id, executionResult, throwable);
                            return null;
                        }), ctx.eventLoop());
```

jrhee17 commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/GraphqlWSSubProtocol.java:
```diff
@@ -0,0 +1,356 @@
+        }
+    }
+
+    private void handleBlocking(WebSocketWriter out, String id,
```
Naming) This isn't handled from a blocking thread which I think can confuse others. What do you think of `handleExecutionResult`?

```suggestion
    private void handleExecutionResult(WebSocketWriter out, String id,
```

jrhee17 reviewed (commented):
Looks almost done! Left some final comments 👍

spacecowboy commented:
Fixed latest round of comments @jrhee17

spacecowboy commented:
Any updates @jrhee17 ?

jrhee17 commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/ExecutionResultSubscriber.java:
```diff
@@ -0,0 +1,114 @@
+
+    @Override
+    public void onNext(ExecutionResult executionResult) {
+        if (subscription == null) {
```
cleanup) I think it is safe to just assert that the `subscriber` is not null since `onSubscribe()` is guaranteed to be called before `onNext`, and `onNext` won't be called after `onComplete()`, `onError()`.

jrhee17 commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/ExecutionResultSubscriber.java:
```diff
@@ -0,0 +1,114 @@
+                }
+            }
+        } catch (JsonProcessingException e) {
+            logger.error("Error serializing graphql result", e);
```
Instead of logging, what do you think of propagating this error to the client so it is more clear which request invoked this error?

```
assert subscription != null;
try {
    if (executionResult.getErrors().isEmpty()) {
        protocol.sendResult(operationId, executionResult);
        requestMore();
    } else {
        protocol.sendGraphqlErrors(executionResult.getErrors());
        subscription.cancel();
    }
} catch (JsonProcessingException e) {
    protocol.completeWithError(e);
    if (subscription != null) {
        subscription.cancel();
    }
}
```

We may need to add a guard to the `ExecutionResultSubscriber` implementation to prevent multiple invocations:

```
new ExecutionResultSubscriber(id, new GraphqlSubProtocol() {

    boolean completed;
...
@Override
public void completeWithError(Throwable cause) {
    if (completed) {
        return;
    }
    completed = true;
...
@Override
public void complete() {
    if (completed) {
        return;
    }
    completed = true;
...
```

jrhee17 commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/GraphqlWSSubProtocol.java:
```diff
@@ -0,0 +1,353 @@
+                    requireNonNull(id, "id");
+                    final ExecutionResultSubscriber s = graphqlSubscriptions.remove(id);
+                    if (s != null) {
+                        s.setCompleted();
```
If a subscription isn't received yet (`ExecutionResultSubscriber.subscription == null`) , I guess a stream may not be completed.

What do you think of modifying `ExecutionResultSubscriber` so that it can subscribe only once?
i.e.
ExecutionResultSubscriber.java
```
@Override
public void onSubscribe(Subscription s) {
    if (subscription != null) {
        s.cancel();
        return;
    }
...
    public void setCompleted() {
        if (subscription == null) {
            subscription = NoopSubscription.get();
        }
        subscription.cancel();
    }
```
This way, a late subscription will still be cancelled

jrhee17 commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/GraphqlServiceBuilder.java:
```diff
@@ -263,6 +269,24 @@ public GraphqlServiceBuilder useBlockingTaskExecutor(boolean useBlockingTaskExec
+     */
+    public GraphqlServiceBuilder webSocketBuilderCustomizer(
+            Consumer<WebSocketServiceBuilder> webSocketBuilderCustomizer) {
+        checkState(useWebSocket, "useWebSocket must be true to customize WebSocketServiceBuilder");
```
Do you mind checking this at `build()` instead? This check may pass if the user sets `useWebSocket == true` afterwards.

```
public GraphqlService build() {
...
checkArgument(useWebSocket || webSocketBuilderCustomizer == null,
                      "useWebSocket must be true to customize WebSocketServiceBuilder");
```

jrhee17 reviewed (commented):
Sorry about the late check, still looks good overall. Left some minor comments I found 🙇

spacecowboy commented:
@jrhee17 I've gone over the comments. They seemed appropriate.

I've also added a BlockHoundIntegration file as discussed.

jrhee17 reviewed (approved):
Changes look good to me 👍   Can you take a look at the lint failures?

jrhee17 commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/GraphqlWSSubProtocol.java:
```diff
@@ -0,0 +1,358 @@
+        try {
+            final String event = serializeToJson(response);
+            logger.trace("NEXT: {}", event);
+            out.write(event);
```
I found that if a connection is closed and a write occurs, an exception can be thrown since the stream `WebSocketWriter` is closed.

```suggestion
            out.tryWrite(event);
```

Can be reproduced by modifying the test:
```
outbound.write("{\"type\":\"connection_init\"}");
outbound.write("{\"type\":\"connection_init\"}");
outbound.write("{\"id\":\"1\",\"type\":\"subscribe\",\"payload\":{\"query\":\"subscription {hello}\"}}");
```

jrhee17 commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/GraphqlWebSocketSubscriber.java:
```diff
@@ -0,0 +1,123 @@
+            case PONG:
+            // Continuation is not mentioned in the spec. Should never happen.
+            case CONTINUATION:
+                logger.trace("Ignoring frame type: {}", webSocketFrame.type());
```
Note to other reviewers: `CONTINUATION` will be handled separately

minwoox replied:
Sent the PR: https://github.com/line/armeria/pull/5357

spacecowboy commented:
Updated @jrhee17 !

minwoox commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/GraphqlWSSubProtocol.java:
```diff
@@ -0,0 +1,358 @@
+
+                    @Override
+                    public ErrorClassification getErrorType() {
+                        return ErrorClassification.errorClassification("Unknown");
```
Question: Shouldn't we use the exception type instead? 🤔

spacecowboy replied:
There is no obvious mapping between a generic `Throwable` and a GraphQL Error classification.

I updated to the built-in generic `ErrorType.DataFetchingException`

minwoox commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/GraphqlWSSubProtocol.java:
```diff
@@ -0,0 +1,358 @@
+    @Nullable
+    public void handle(String event, WebSocketWriter out) {
+        try {
+            final Map<String, Object> eventMap = parseJsonString(event, JSON_MAP);
+            final String type = toStringFromJson("type", eventMap.get("type"));
+            requireNonNull(type, "type");
```
How about sending the close with 4400 status when an exception is raised while parsing the event?
https://github.com/enisdenjo/graphql-ws/blob/master/PROTOCOL.md#invalid-message
I think we also have to call `subscription.cancel` when it fails to process.

In order to do that, how about just merging `GraphqlWebSocketSubscriber` and `GraphqlWSSubProtocol` into one?
`GraphqlWebSocketSubscriber` doesn't have much logic in it so I think we can merge those.

spacecowboy replied:
Personally, I'd prefer not. `graphql-ws` is not the first protocol, and I doubt it will be the last.

About the 4400 status - we are closing the websocket on such a failure. So it's not obvious to me that the client would ever have the chance to handle the GraphQL error before the websocket closes on them.

minwoox commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/GraphqlWSSubProtocol.java:
```diff
@@ -0,0 +1,358 @@
+                case "connection_init":
+                    if (connectionInitiated) {
+                        // Already initiated, that's an error
+                        throw new Exception("4429: Too many initialisation requests");
```
I think this isn't going to work as expected. We need to send the close status explicitly.

spacecowboy replied:
This exception will close the websocket on the server end. The text is just descriptive.

minwoox commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/GraphqlWSSubProtocol.java:
```diff
@@ -0,0 +1,358 @@
+        }
+    }
+
+    private String serializeToJson(Object object) throws JsonProcessingException {
```
Could you add `static` to the methods in this class where applicable?

minwoox reviewed (commented):
Looks great, @spacecowboy! 👍 
Left small suggestions mostly about style issues. 😉

minwoox reviewed (commented):
Looks great, @spacecowboy! 👍 
Left small suggestions mostly about style issues. 😉

spacecowboy commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/GraphqlWebSocketSubscriber.java:
```diff
@@ -0,0 +1,123 @@
+
+    @Override
+    public void onSubscribe(Subscription s) {
+        if (subscription != null) {
+            /*
+            A Subscriber MUST call Subscription.cancel() on the given Subscription after an
+            onSubscribe signal if it already has an active Subscription.
+             */
+            subscription.cancel();
+        }
+        subscription = s;
```
I don't agree. The spec is clear about what needs to happen as can be seen in the comment.

minwoox replied:
That is true.
But if it's obvious that `onSubscribe` is called only once, we don't implement the spec fully.
This is one of the examples:
https://github.com/line/armeria/blob/dc2bdc6107bb1c6784773781cac60cac6ca39203/core/src/main/java/com/linecorp/armeria/client/AbstractHttpRequestSubscriber.java#L72-L74

Sometimes, it's also helpful for the code readers to notice the fact that this subscriber is subscribed only once by implementing this way.

spacecowboy commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/GraphqlWebSocketSubscriber.java:
```diff
@@ -0,0 +1,123 @@
+            return;
+        }
+        switch (webSocketFrame.type()) {
+            // If Binary, handle it same as a text frame. Should never happen.
```
I'll leave the choice to you. It's not clear what to do:

* Do nothing
* Return exception
* Return exception and close subscription

minwoox replied:
I'm leaning toward sending `4400` because we can consider the frame as an invalid message. WDYT?

minwoox replied:
I discovered that the upstream sends 4400 when it's in binary format, as shown in https://github.com/enisdenjo/graphql-ws/blob/812129dcd6454ac7850a14b300162977d82e9bf9/src/server.ts#L590-L595.

They also discussed supporting both text and binary WebSocket frames but they only support text frame for now.
https://github.com/enisdenjo/graphql-ws/issues/409#issuecomment-1271315726

spacecowboy replied:
I think the right course of action is closing the websocket with `INVALID_MESSAGE_TYPE` which explicitly mentions this as an example use case

spacecowboy commented:
@minwoox addressed some of your comments.

minwoox commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/DefaultGraphqlService.java:
```diff
@@ -55,15 +69,49 @@ final class DefaultGraphqlService extends AbstractGraphqlService implements Grap
+
+    @Override
+    public ExchangeType exchangeType(RoutingContext routingContext) {
+        if (isHttp2WebSocketUpgradeRequest(routingContext.headers()) && webSocketService != null) {
```
Created. 😉 
https://github.com/line/armeria/pull/5353

spacecowboy commented:
@minwoox I've addressed the last of the comments (I think).
I added some extra exception handling in order to set the web socket close code correctly.

minwoox commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/GraphqlWSSubProtocol.java:
```diff
@@ -0,0 +1,386 @@
+                            handleExecutionResult(out, id, executionResult, throwable);
+                            return null;
+                        }, ctx.eventLoop());
+                    } catch (WebSocketCloseException e) {
+                        logger.debug("Error handling subscription", e);
+                        // Also cancel subscription if present before closing websocket
+                        final ExecutionResultSubscriber s = graphqlSubscriptions.remove(id);
+                        if (s != null) {
+                            s.setCompleted();
+                        }
+                        out.close(e.getWebSocketCloseStatus());
```
I think we have to cancel all subscriptions instead of just one because we cannot send the message anyway.
So what do you think of removing this catch clause and handle `WebSocketCloseException` at line 179?
We might want to add a method (e.g. `cleanup()`) and call there:

```java
private void cleanup() {
    for (ExecutionResultSubscriber executionResultSubscriber : graphqlSubscriptions.values()) {
        executionResultSubscriber.setCompleted();
    }
}
```

ikhoon replied:
`WebSocketCloseException` is not an IOException. Should we send messages after that?

minwoox commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/GraphqlWSSubProtocol.java:
```diff
@@ -0,0 +1,386 @@
+                    }
+                    final Map<String, Object> payload = toMapFromJson(eventMap.get("payload"));
+                    try {
+                        if (graphqlSubscriptions.containsKey(id)) {
```
Note: Because we don't put the ID into the map right away, there's a chance that multiple operations for the same ID can be received without an error.
However, it's not critical, let's handle it in another PR. 😉

minwoox commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/GraphqlWSSubProtocol.java:
```diff
@@ -0,0 +1,386 @@
+            final String event = serializeToJson(response);
+            logger.trace("NEXT: {}", event);
+            out.tryWrite(event);
+        } catch (JsonProcessingException e) {
+            logger.error("Error serializing next event", e);
+            writeError(out, operationId, e);
+        }
```
It seems like we shouldn't catch the exception here but propagate to the caller so that `subscription.cancel()` is called.
Or, how about renaming this to `tryWriteNext` and returning false when it fails? If so the caller can cancel the subscription.

ikhoon replied:
Changed to propagate the exception to the caller.

minwoox commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/GraphqlWSSubProtocol.java:
```diff
@@ -0,0 +1,386 @@
+        out.tryWrite("{\"type\":\"connection_ack\"}");
+    }
+
+    private void writeNext(WebSocketWriter out, String operationId, ExecutionResult executionResult) {
```
nit:
```suggestion
    private static void writeNext(WebSocketWriter out, String operationId, ExecutionResult executionResult) {
```

minwoox commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/GraphqlWSSubProtocol.java:
```diff
@@ -0,0 +1,386 @@
+            final String event = serializeToJson(errorResponse);
+            logger.trace("ERROR: {}", event);
+            out.tryWrite(event);
+        } catch (JsonProcessingException e) {
+            logger.error("Error serializing error event", e);
+            writeError(out, operationId, e);
```
ditto

minwoox reviewed (commented):
I think we are pretty close to the end.
I left a few more suggestions. PTAL. 🙇

ikhoon commented on graphql-protocol/src/main/java/com/linecorp/armeria/server/graphql/protocol/AbstractGraphqlService.java:
```diff
@@ -63,8 +66,20 @@ public abstract class AbstractGraphqlService extends AbstractHttpService {
 
+    @Override
+    protected HttpResponse doConnect(ServiceRequestContext ctx, HttpRequest req) throws Exception {
+        if (ctx.sessionProtocol().isExplicitHttp2() && isHttp2WebSocketUpgradeRequest(req.headers())) {
```
Sorry for the late review. 

There are two issues with the current approach.

1. `DefaultWebSocketService` is exposed as a public API. I understood why it was necessary. There was no extension point on `WebSocketService` API design.

2. The WebSocket handshake is a bit hard-coded in `DefaultGraphqlService`. We may delegate the handshake to `WebSocketService` and implement two methods. One is for WebSocket and another is for normal HTTP requests.

Proposal: 

Make `WebSocketService` abstract class so that `GraphqlService` can support WebSocket by extending or composing it.

```java
abstract class WebSocketService {

   protected abstract WebSocket handleWebSocket(ServiceRequestContext ctx, WebSocket in);

   protected HttpResponse handleHttp(ServiceRequestContext ctx, HttpRequest req) {
        return HttpResponse.of(HttpStatus.BAD_REQUEST, MediaType.PLAIN_TEXT_UTF_8,
                               "The upgrade header must contain");
   }
}
```
By changing the design of `WebSocketService`
1) We may split WebSocket logic from GraphQL over HTTP.
```java
class GraphqlWebSocketService extends WebSocketService implements GraphqlService {

  private final GraphqlService GraphqlService graphqlService;
   
  @Override
  protected WebSocket handleWebSocket(ServiceRequestContext ctx, WebSocket in) {
    // execute Graphql over WebSocket
  }

  @Override
  protected HttpResponse handleHttp(ServiceRequestContext ctx, HttpRequest req) {
    // delegate to the original GraphqlService
    return graphqlService.serve(ctx, req);
  }
}
```
2) Two GraphQL services can only be composed when WebSocket is enabled.
```java
class GraphqlServiceBuilder {

  GraphqlService build() {
    DefaultGraphqlService graphqlService = ...;
    if (enableWebSocket) {
       return new GraphqlWebSocketService(graphqlService);
    } else {
      return graphqlService;
    }
  }
}
```

If this proposal is acceptable, let me send a PR to refactor `WebSocketService` first.

jrhee17 replied:
> DefaultWebSocketService is exposed as a public API. I understood why it was necessary. There was no extension point on WebSocketService API design.

I don't think this is necessarily due to a design choice. For instance, I think it is trivial to modify the current `DefaultWebSocketService` to be package-private in this pull request

> The WebSocket handshake is a bit hard-coded in DefaultGraphqlService. We may delegate the handshake to WebSocketService and implement two methods. One is for WebSocket and another is for normal HTTP requests.

I agree with this sentiment

ikhoon replied:
After discussion with other maintainers, I updated the API design to easily compose `WebSocketService`. #5368

spacecowboy replied:
Nice. So with #5368 merged, if I understand it correctly, it would be possible to implement this entire PR by any user of the API without modifying Armeria itself?

Making this PR technically unnecessary, but just still a nice-to-have?

ikhoon replied:
> it would be possible to implement this entire PR by any user of the API without modifying Armeria itself?

Right. The new API in `WebSocketService` also allows to set a fallback `HttpService` for non-WebSocket requests. That means we can serve `WebSocketService` and `HttpService` at the same path.

> Making this PR technically unnecessary, but just still a nice-to-have?

We don't need to check WebSocket upgrade in `DefaultGraphqlService directly` but I believe `GraphqlWebSocketSubscriber` and related classes are still worth adding.

If #5368 is merged, GraphQL over WebSocket will be supported by the following steps:
```java
class GraphqlServiceBuilder {
   GraphqlService build() {
    // DefaultGraphqlService is used as is without modifying
    var graphqlService = ...
    if (webSocketEnabled) {
      val webSocketService = 
      WebSocketService
        .builder(graphqlWebSocketHandler)
        .fallbackService(graphqlService)
        ...
        .build()
      graphqlService = new GraphqlWebSocketService(webSocketService)
    }
    return graphqlService;
  }
}
```

`GraphqlWebSocketService` will be a simple wrapper that implements both `GraphqlService` and `WebSocketService`.
```java
class GraphqlWebSocketService implements GraphqlService, WebSocketService {
   WebSocketService delegate;

   @Override
   public WebSocket serve(ServiceRequestContext ctx, WebSocket in) {
     return delegate.serve(ctx, in);
   }
   ...
}
```

jrhee17 replied:
> class GraphqlWebSocketService implements GraphqlService

Note: I prefer that `GraphqlWebSocketService` does not inherit `WebSocketService` so that HTTP requests do not unintentionally have larger resource limitations than intended. (or at least there is some way to ignore such services)

ref: https://github.com/line/armeria/blob/067578f161d43cc981b3b5c97f9fd385131c77cd/core/src/main/java/com/linecorp/armeria/server/ServiceConfigBuilder.java#L324-L351

ikhoon replied:
Good point. From an implementation perspective, it doesn't seem to be a problem that `GraphqlWebSocketService` inherits `WebSocketService`. The genuine problem is that the configuration is hard-coded with the `WebSocketService.class`.

In order not to set an unexpected configuration to `Graphql(WebSocket)Service`, we need to change the code to apply the default WebSocket settings only for `DefaultWebSocketService`.

Had a chat with other folks and we decided to provide dynamic options for a service. `RequestOptions` is going to be added to `HttpService`.
```java
interface HttpService {
   default RequestOptions requestOptions(RoutingContext routingContext) {
     return RequestOptions.of();
   }
}

interface WebSocketService {
   default RequestOptions requestOptions(RoutingContext routingContext) {
     return new WebRequestOptions(...);
   }
}

class GraphqlWebSocketService {
   default RequestOptions requestOptions(RoutingContext routingContext) {
      if (isWebSocket(routingContext) { 
        return new WebRequestOptions(...);
      } else {
        return RequestOptions.of();
      }
   }
}
```

ikhoon commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/WebSocketCloseException.java:
```diff
@@ -0,0 +1,37 @@
+
+import com.linecorp.armeria.common.websocket.WebSocketCloseStatus;
+
+public class WebSocketCloseException extends Exception {
```
The exception is only used to propagate WebSocket close status to other implementations and is not exposed to the userland.
Instead, let me remove `public` visibility.

ikhoon commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/GraphQlBlockHoundIntegration.java:
```diff
@@ -0,0 +1,30 @@
+/**
+ * A {@link BlockHoundIntegration} for the graphql module.
+ */
+public class GraphQlBlockHoundIntegration implements BlockHoundIntegration {
```
This class has been removed because `ExecutionId.generate()` is no longer used.

ikhoon commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/GraphqlServiceBuilder.java:
```diff
@@ -263,6 +270,23 @@ public GraphqlServiceBuilder useBlockingTaskExecutor(boolean useBlockingTaskExec
+     * Sets an optional {@link WebSocketServiceBuilder} customizer.
+     */
+    public GraphqlServiceBuilder webSocketBuilderCustomizer(
+            Consumer<WebSocketServiceBuilder> webSocketBuilderCustomizer) {
```
I haven't addressed this comment in this PR. Considering the size of this PR, what do you think of handling this issue in a separate PR?

ikhoon commented:
The new `WebSocketService` was applied to `GraphqlWebSocketService`. All changes in `DefaultGraphqlService` were reverted. PTAL. 🙇‍♂️

jrhee17 commented:
> PTAL

Question) Is this PR ready for review? Is there no need to also merge https://github.com/line/armeria/pull/5357 and https://github.com/line/armeria/pull/5353, or can these be dealt with separately?

ikhoon commented:
- #5353 Let's handle separately
- #5357 It seems good to enable `aggregateContinuation` in this PR.

ikhoon commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/GraphqlServiceBuilder.java:
```diff
@@ -263,6 +270,24 @@ public GraphqlServiceBuilder useBlockingTaskExecutor(boolean useBlockingTaskExec
     }
 
+    /**
+     * Enable <a href="https://github.com/enisdenjo/graphql-ws/blob/master/PROTOCOL.md">GraphQL over WebSocket Protocol</a>.
```
```suggestion
     * Enables <a href="https://github.com/enisdenjo/graphql-ws/blob/master/PROTOCOL.md">GraphQL over WebSocket Protocol</a>.
```

minwoox reviewed (approved):
Thanks, @spacecowboy for adding this great feature! 🎉 🎉 🎉 
Thanks, @ikhoon for cleaning this up! 🎉 🎉 🎉

ikhoon reviewed (approved):
Fantastic work! Thanks, @spacecowboy 🙇‍♂️🚀
